# Data understanding

#### a) Define the question

Perform classification of the testing set samples using the Naive Bayes Classifier on the data and compute the accuracy.

#### b) Metrics for success

The project will be considered successful if we are able to get a high accuracy score after comparing the different test sizes, with the Optimal Naive Bayes Classifier.

#### c) Recording the experimental design 

The following steps will be followed during this exercise:
- Data Understanding
- Data Preparation
- Data Cleaning
- Multivariate Analysis
- Modelling with Naive Bayes Classifier
- Evaluation
- Challenging the solution

#### d) Data relevance. 

This will be described in our data exploration

# Data preparation

## Importing the libraries. 

In [62]:
# Import all necessary libraries. 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn import metrics
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.model_selection import RepeatedStratifiedKFold
import xgboost as xgb
from xgboost import XGBClassifier




## Reading the Data 

In [3]:
# Loading our datset
#
df = pd.read_csv('spambase.data')


In [4]:
# Previewing the first 10 records of our dataframe
#
df.head(10)

0  0.64  0.64.1  0.1  0.32   0.2  ...   0.43   0.44  3.756   61   278  1
0  0.21  0.28    0.50  0.0  0.14  0.28  ...  0.180  0.048  5.114  101  1028  1
1  0.06  0.00    0.71  0.0  1.23  0.19  ...  0.184  0.010  9.821  485  2259  1
2  0.00  0.00    0.00  0.0  0.63  0.00  ...  0.000  0.000  3.537   40   191  1
3  0.00  0.00    0.00  0.0  0.63  0.00  ...  0.000  0.000  3.537   40   191  1
4  0.00  0.00    0.00  0.0  1.85  0.00  ...  0.000  0.000  3.000   15    54  1
5  0.00  0.00    0.00  0.0  1.92  0.00  ...  0.054  0.000  1.671    4   112  1
6  0.00  0.00    0.00  0.0  1.88  0.00  ...  0.000  0.000  2.450   11    49  1
7  0.15  0.00    0.46  0.0  0.61  0.00  ...  0.203  0.022  9.744  445  1257  1
8  0.06  0.12    0.77  0.0  0.19  0.32  ...  0.081  0.000  1.729   43   749  1
9  0.00  0.00    0.00  0.0  0.00  0.00  ...  0.000  0.000  1.312    6    21  1

[10 rows x 58 columns]

In [5]:
# Checking the data types of our different columns
#
spamdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4600 entries, 0 to 4599
Data columns (total 58 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       4600 non-null   float64
 1   0.64    4600 non-null   float64
 2   0.64.1  4600 non-null   float64
 3   0.1     4600 non-null   float64
 4   0.32    4600 non-null   float64
 5   0.2     4600 non-null   float64
 6   0.3     4600 non-null   float64
 7   0.4     4600 non-null   float64
 8   0.5     4600 non-null   float64
 9   0.6     4600 non-null   float64
 10  0.7     4600 non-null   float64
 11  0.64.2  4600 non-null   float64
 12  0.8     4600 non-null   float64
 13  0.9     4600 non-null   float64
 14  0.10    4600 non-null   float64
 15  0.32.1  4600 non-null   float64
 16  0.11    4600 non-null   float64
 17  1.29    4600 non-null   float64
 18  1.93    4600 non-null   float64
 19  0.12    4600 non-null   float64
 20  0.96    4600 non-null   float64
 21  0.13    4600 non-null   float64
 22  

> All the columns in the dataset seem to have the appropriate datatype. 

In [6]:
df.describe()

0         0.64  ...           278            1
count  4600.000000  4600.000000  ...   4600.000000  4600.000000
mean      0.104576     0.212922  ...    283.290435     0.393913
std       0.305387     1.290700  ...    606.413764     0.488669
min       0.000000     0.000000  ...      1.000000     0.000000
25%       0.000000     0.000000  ...     35.000000     0.000000
50%       0.000000     0.000000  ...     95.000000     0.000000
75%       0.000000     0.000000  ...    265.250000     1.000000
max       4.540000    14.280000  ...  15841.000000     1.000000

[8 rows x 58 columns]

> These are the summary statistics of our dataset

## Exploring the Data

In [7]:
# Checking the number of rows and columns
#
print(df.shape)

(4600, 58)


> Our Spam feature dataset has 58 columns and 4600 entries. 

In [8]:
# Renaming our dataset as per the description provided
# 
column_names = ['word_freq_make'  ,'word_freq_address' ,'word_freq_all' ,'word_freq_3d' ,'word_freq_our'        
,'word_freq_over' ,'word_freq_remove' ,'word_freq_internet' ,'word_freq_order' ,'word_freq_mail','word_freq_receive' ,'word_freq,_will' ,'word_freq_people' ,'word_freq_report' ,'word_freq_addresses'  
,'word_freq_free' ,'word_freq_business' ,'word_freq_email' ,'word_freq_you' ,'word_freq_credit'     
,'word_freq_your' ,'word_freq_font' ,'word_freq_000' ,'word_freq_money' ,'word_freq_hp' ,'word_freq_hpl' ,'word_freq_george' ,'word_freq_650' ,'word_freq_lab' ,'word_freq_labs'       
,'word_freq_telnet' ,'word_freq_857' ,'word_freq_data' ,'word_freq_415' ,'word_freq_85' ,'word_freq_technology' 
,'word_freq_1999' ,'word_freq_parts' ,'word_freq_pm' ,'word_freq_direct' ,'word_freq_cs' ,'word_freq_meeting' ,'word_freq_original' ,'word_freq_project' ,'word_freq_re' ,'word_freq_edu' ,'word_freq_table' ,'word_freq_conference' 
,'char_freq_;' ,'char_freq_(' ,'char_freq_[' ,'char_freq_!' ,'char_freq_$' ,'char_freq_#','capital_run_length_average'
,'capital_run_length_longest' ,'capital_run_length_total', 'class']

df.columns = column_names
df.head()

word_freq_make  word_freq_address  ...  capital_run_length_total  class
0            0.21               0.28  ...                      1028      1
1            0.06               0.00  ...                      2259      1
2            0.00               0.00  ...                       191      1
3            0.00               0.00  ...                       191      1
4            0.00               0.00  ...                        54      1

[5 rows x 58 columns]

# Data Cleaning


In [9]:
# Cheking our column names
#
df.columns

Index(['word_freq_make', 'word_freq_address', 'word_freq_all', 'word_freq_3d',
       'word_freq_our', 'word_freq_over', 'word_freq_remove',
       'word_freq_internet', 'word_freq_order', 'word_freq_mail',
       'word_freq_receive', 'word_freq,_will', 'word_freq_people',
       'word_freq_report', 'word_freq_addresses', 'word_freq_free',
       'word_freq_business', 'word_freq_email', 'word_freq_you',
       'word_freq_credit', 'word_freq_your', 'word_freq_font', 'word_freq_000',
       'word_freq_money', 'word_freq_hp', 'word_freq_hpl', 'word_freq_george',
       'word_freq_650', 'word_freq_lab', 'word_freq_labs', 'word_freq_telnet',
       'word_freq_857', 'word_freq_data', 'word_freq_415', 'word_freq_85',
       'word_freq_technology', 'word_freq_1999', 'word_freq_parts',
       'word_freq_pm', 'word_freq_direct', 'word_freq_cs', 'word_freq_meeting',
       'word_freq_original', 'word_freq_project', 'word_freq_re',
       'word_freq_edu', 'word_freq_table', 'word_freq_conference',

> Our column titles have no issues and are all consistent.

In [10]:
# Checking for the missing values
#
df.isna().sum()

word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_our                 0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq,_will               0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet              0
word_fre

> We have no missing data. 

# Multivariate Analysis

In [15]:
# spliting the data into X and y
X = df.drop('class', 1)
y = df['class'].values


In [16]:
from sklearn.model_selection import train_test_split
# Split into training and testing data 

X_train, X_test, y_train, y_test = train_test_split(X, y , test_size = 0.2, random_state = 42)

In [ ]:
# feature scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)


In [17]:
# performing LDA
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
lda = LDA(n_components = 13)
X_train = lda.fit_transform(X_train, y_train)
X_test = lda.transform(X_test)


/usr/local/lib/python3.7/dist-packages/sklearn/discriminant_analysis.py:463: ChangedBehaviorWarning: n_components cannot be larger than min(n_features, n_classes - 1). Using min(n_features, n_classes - 1) = min(57, 2 - 1) = 1 components.
  ChangedBehaviorWarning)
/usr/local/lib/python3.7/dist-packages/sklearn/discriminant_analysis.py:469: FutureWarning: In version 0.23, setting n_components > min(n_features, n_classes - 1) will raise a ValueError. You should set n_components to None (default), or a value smaller or equal to min(n_features, n_classes - 1).
  warnings.warn(future_msg, FutureWarning)


In [18]:
# Getting the LDA coeficients

factors = pd.DataFrame (index = X.columns.values, data = lda.coef_[0].T)
factors.sort_values(0, ascending = False).head(10)

0
word_freq_remove    2.122862
char_freq_$         2.022548
word_freq_000       1.858357
word_freq_over      1.096539
word_freq_internet  0.983911
word_freq_money     0.882474
word_freq_order     0.822565
word_freq_our       0.805913
word_freq_free      0.760607
word_freq_credit    0.681850

> From this we can see that our most important features are word_freq_remove then char_freq_$, word_freq_000, word_freq_over, word_freq_internet, word_freq_money, word_freq_order, word_freq_our, word_freq_free then word_freq_credit respectively in that order.

# Modelling 

## Baseline Modelling

In [45]:
# Defining our predictor and target variables
#
X = df.drop(['class'], axis = 1)
y = df['class']

# Spliting our dataset
#
X_train, X_test, y_train, y_test = train_test_split(X,y, random_state = 0, test_size = 0.2)
# Scaling predictor variables
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.fit_transform(X_test)

# Fitting the data
regressor = LogisticRegression()
regressor.fit(X_train, y_train)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [23]:
# Making the prediction. 
#
y_pred = regressor.predict(X_test)

In [26]:
# Getting the score of the baseline model. 
#
cm = confusion_matrix(y_test, y_pred)
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))  
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))
print('The confusuion matrix is: ', "\n",cm)
print ('The accuracy score: ', accuracy_score(y_test, y_pred))

Root Mean Squared Error: 0.2553769592276246
Mean Squared Error: 0.06521739130434782
The confusuion matrix is:  
 [[513  25]
 [ 35 347]]
The accuracy score:  0.9347826086956522


> Our model has an RMSE of 0.2553 which is sufficiently low, showing that our model is a good fit. The rest of the metrice are also displayed above, and they support that the logistic regression model is a good fit.  

## Naive Bayes Model
We shall use the Gaussian Naive Bayes model since our data set is continuous. We shall assume that our features follow a normal distribution.

In [29]:
# Fitting our model
#
clf = GaussianNB()  
model = clf.fit(X_train, y_train) 

In [36]:
# Predicting our test predictors
predicted = model.predict(X_test)
print ('The accuracy score: ', accuracy_score(y_test, predicted))

The accuracy score:  0.8402173913043478


> Using Gaussian Naive Bayes, we have an accuracy score of 84% which generally is a good fit, but not high enough for us.

### Hyperparameter tuning

In [39]:
cv_method = RepeatedStratifiedKFold(n_splits=5,  n_repeats=3, random_state=999)

In [41]:
from sklearn.model_selection import GridSearchCV
params_NB = {'var_smoothing': np.logspace(0,-7, num=100)}
gs_NB = GridSearchCV(estimator=model, param_grid=params_NB, cv=cv_method,verbose=1,scoring='accuracy')
gs_NB.fit(X_test, y_test);

Fitting 15 folds for each of 100 candidates, totalling 1500 fits


[Parallel(n_jobs=1)]: Using backend SequentialBackend with 1 concurrent workers.
[Parallel(n_jobs=1)]: Done 1500 out of 1500 | elapsed:    3.4s finished


In [42]:
# Predict
predict_test = gs_NB.predict(X_test)
# Check accuracy Score on test dataset
accuracy_test = accuracy_score(y_test,predict_test)
print('accuracy_score on test dataset : ', accuracy_test)

accuracy_score on test dataset :  0.8369565217391305


> Our tuned model has a lower accuracy score than our untuned model, thus we shall go with the untuned model 

In [46]:
# Defining the list of test sizes we will use for the assessment
#
test_size = [0.1, 0.2, 0.3, 0.4, 0.5]

# Using a for loop to split the dataset, fit it with the optimal parameters then get the accuracy score.
#
for test in test_size:
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = test, random_state = 0)
    
    # Fitting to the classifier
    clf = GaussianNB()  
    model = clf.fit(X_train, y_train)
    # Predicting our test predictors
    predicted = model.predict(X_test)

    print("Test size {} has accuracy score:".format(test), (metrics.accuracy_score(y_test, predicted)*100))

Test size 0.1 has accuracy score: 82.3913043478261
Test size 0.2 has accuracy score: 83.04347826086956
Test size 0.3 has accuracy score: 82.82608695652173
Test size 0.4 has accuracy score: 82.01086956521739
Test size 0.5 has accuracy score: 81.91304347826087


> The model with a test size of 0.2 has the highest accuracy score, thus is the most optimal test size.

# Challenging the Solution
We will challenge the solution with XGBoost algorithm.

> 

In [55]:
# Defining the regression of our classifer to linear
#
xgb_model = xgb.XGBClassifier(objective='reg:linear',random_state=42)

In [70]:
# Defining the predictor and target variables
#
import re
regex = re.compile(r"\[|\]|<", re.IGNORECASE)
df.columns = [regex.sub("_", col) if any(x in str(col) for x in set(('[', ']', '<'))) else col for col in df.columns.values]


X = df.drop(['class'], axis = 1)
y = df['class']

# Splitting the data
#
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size = 0.2, random_state = 42)
# Fitting the data into our model
#

xgb_model.fit(X_train, y_train)

[19:06:03] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='reg:linear', random_state=42,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [71]:
# Predict the data using the test set
#
y_predict = xgb_model.predict(X_test)

In [73]:
# Printing out the various metrics
#
from sklearn.metrics import classification_report
print(confusion_matrix(y_test,y_predict))
print(classification_report(y_test,y_predict))

[[507  23]
 [ 53 337]]
              precision    recall  f1-score   support

           0       0.91      0.96      0.93       530
           1       0.94      0.86      0.90       390

    accuracy                           0.92       920
   macro avg       0.92      0.91      0.91       920
weighted avg       0.92      0.92      0.92       920



> Our XGboost model has an accuracy score of 92% and an f1-score of 93%, which is  better than the metrics we got for Gaussian Naive Bayes model, thus it's a better model for our dataset